In [88]:
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime
from pandas import Series
import warnings

In [89]:
#importamos archivos
entrenamiento = pd.read_csv("Train_SU63ISt.csv")
prueba = pd.read_csv("Test_0qrQsBz.csv")
#Creamos copia 
entrenamiento_original = entrenamiento.copy()
prueba_original = prueba.copy()

In [90]:
#Características de datos
entrenamiento.columns, prueba.columns

(Index(['ID', 'Datetime', 'Count'], dtype='object'),
 Index(['ID', 'Datetime'], dtype='object'))

In [91]:
#Característica de cada columna
entrenamiento.dtypes, prueba.dtypes

(ID           int64
 Datetime    object
 Count        int64
 dtype: object,
 ID           int64
 Datetime    object
 dtype: object)

In [92]:
#Observamos las dimensiones
entrenamiento.shape, prueba.shape

((18288, 3), (5112, 2))

Tenemos 18288 y 5112 filas, 3 y 2 columnas para los conjuntos de entrenamiento y prueba respectivamente. 

A continuación extraeremos la información para encontrar las relaciones en nuestro conjunto de datos. Como se verificó anteriormente nuestra columna "Datetime" es un objeto por lo que que tendremos que cambiar el tipo de dato para poder extraer información de esta columna.

In [93]:
#Transformamos la columna Datetime a formate date time
entrenamiento["Datetime"] = pd.to_datetime(entrenamiento.Datetime, format="%d/%m/%Y %H:%M")
prueba["Datetime"] = pd.to_datetime(prueba.Datetime, format="%d/%m/%Y %H:%M")
entrenamiento_original["Datetime"] = pd.to_datetime(entrenamiento_original.Datetime, format="%d/%m/%Y %H:%M")
prueba_original["Datetime"] = pd.to_datetime(prueba_original.Datetime, format="%d/%m/%Y %H:%M")

In [94]:
#Extraemos el años, mes, día y hora 

In [95]:
for i in (entrenamiento, prueba, entrenamiento_original, prueba_original):
    i["año"]=i.Datetime.dt.year
    i["mes"]=i.Datetime.dt.month
    i["dia"]=i.Datetime.dt.day
    i["hora"]=i.Datetime.dt.hour

# Hipótesis

Si observamos los datos podemos darnos cuenta que con el paso de los años aumenta el flujo de personas. Además durante un año tendremos flujos más altos en los meses de mayo y octubre, asimismo durante los días de la semana tendremos más gente en los fines de semana y días festivos, también podemos identificar las horas pico.

In [80]:
#Observamos 10 primeros datos
entrenamiento[:20] #observar primeros 20 datos

ID            Datetime  Count   año  mes  dia  hora
0    0 2012-08-25 00:00:00      8  2012    8   25     0
1    1 2012-08-25 01:00:00      2  2012    8   25     1
2    2 2012-08-25 02:00:00      6  2012    8   25     2
3    3 2012-08-25 03:00:00      2  2012    8   25     3
4    4 2012-08-25 04:00:00      2  2012    8   25     4
5    5 2012-08-25 05:00:00      2  2012    8   25     5
6    6 2012-08-25 06:00:00      2  2012    8   25     6
7    7 2012-08-25 07:00:00      2  2012    8   25     7
8    8 2012-08-25 08:00:00      6  2012    8   25     8
9    9 2012-08-25 09:00:00      2  2012    8   25     9
10  10 2012-08-25 10:00:00      2  2012    8   25    10
11  11 2012-08-25 11:00:00      6  2012    8   25    11
12  12 2012-08-25 12:00:00      4  2012    8   25    12
13  13 2012-08-25 13:00:00      2  2012    8   25    13
14  14 2012-08-25 14:00:00      6  2012    8   25    14
15  15 2012-08-25 15:00:00      2  2012    8   25    15
16  16 2012-08-25 16:00:00      2  2012    8   25    16
17  17 2012-08-25 17:00:00      2  2012    8   25    17
18  18 2012-08-25 18:00:00      2  2012    8   25    18
19  19 2012-08-25 19:00:00      2  2012    8   25    19

Anteriormente supusimos que hay un mayor flujo en el fin de semana para verificar dicha hipótesis crearemos una nueva variable para asignar los días fin de semana. En este caso observando los datos podemos deducir que los sábodos y domingo corresponden a los días 5 y 6.

In [84]:
entrenamiento["dia semana"] = entrenamiento["Datetime"].dt.dayofweek
temporal = entrenamiento["Datetime"]

Ahora verificaremos si efectivamente es sábado o domingo

In [83]:
def finde(fila):
    if fila.dayofweek == 5 or fila.dayofweek == 6:
        return 1
    else:
        return 0

In [87]:
temporal2 = entrenamiento["Datetime"].apply(finde)
entrenamiento["Fin de semana"] = temporal2